src/driver/odoc_driver.ml

```diff
@@ -564,26 +564,41 @@ let run libs verbose packages_dir odoc_dir odocl_dir html_dir stats nb_workers
   let () = Worker_pool.start_workers env sw nb_workers in
 
   let all, extra_libs_paths =
-    match (voodoo, package_name, dune_style, packages_dir) with
-    | true, Some p, None, None ->
+    match (voodoo, package_name, dune_style, packages_dir, libs, packages) with
+    | true, Some p, None, None, [], [] ->
         let all = Voodoo.of_voodoo p ~blessed in
         let extra_libs_paths = Voodoo.extra_libs_paths odoc_dir in
         (all, extra_libs_paths)
-    | false, None, Some dir, None ->
+    | false, None, Some dir, None, [], [] ->
         (Dune_style.of_dune_build dir, Util.StringMap.empty)
-    | false, None, None, packages_dir ->
-        let libs = if libs = [] then Ocamlfind.all () else libs in
+    | false, None, None, packages_dir, [], [] ->
+        let libs = Ocamlfind.all () in
         let libs =
           List.map Ocamlfind.sub_libraries libs
           |> List.fold_left Util.StringSet.union Util.StringSet.empty
         in
         (Packages.of_libs ~packages_dir libs, Util.StringMap.empty)
-    | true, None, _, _ -> failwith "--voodoo requires --package"
-    | false, Some _, _, _ -> failwith "--package requires --voodoo"
-    | true, _, _, Some _ | false, _, Some _, Some _ ->
+    | false, None, None, packages_dir, libs, [] ->
+        let libs =
+          List.map Ocamlfind.sub_libraries libs
+          |> List.fold_left Util.StringSet.union Util.StringSet.empty
+        in
+        (Packages.of_libs ~packages_dir libs, Util.StringMap.empty)
+    | false, None, None, packages_dir, [], packages ->
+        (Packages.of_packages ~packages_dir packages, Util.StringMap.empty)
+    | true, Some _p, None, None, _, _ ->
+        failwith "--voodoo and -l / -p are mutually exclusive"
+    | false, None, Some _dir, None, _, _ ->
+        failwith "--dune-style and -l / -p are mutually exclusive"
+    | true, None, _, _, _, _ -> failwith "--voodoo requires --package"
+    | false, Some _, _, _, _, _ -> failwith "--package requires --voodoo"
+    | true, _, _, Some _, _, _ | false, _, Some _, Some _, _, _ ->
         failwith "--packages-dir is only useful in opam mode"
-    | true, _, Some _, _ ->
+    | true, _, Some _, _, _, _ ->
         failwith "--voodoo and --dune-style are mutually exclusive"
+    | false, None, None, _packages_dir, _, _ ->
+        failwith
+          "Please specify either packages (-p) or libraries (-l), not both"
```

Comment:
This was bad before, now it's terrible. We should split this into multiple subcommands

Reply:
I opened #1237